rmw_iceoryx_cpp/src/rmw_service.cpp

```diff
@@ -34,26 +39,94 @@ rmw_create_service(
   rmw_service_t * rmw_service = nullptr;
+  iox::popo::UntypedServer * iceoryx_server = nullptr;
+  IceoryxServer * iceoryx_server_abstraction = nullptr;
+
+  bool returnOnError = false;
+
+  auto cleanupAfterError = [&]() {
+      if (rmw_service) {
+        if (iceoryx_server) {
+          RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+            iceoryx_server->~UntypedServer(), iox::popo::UntypedServer)
+          rmw_free(iceoryx_server);
+        }
+        if (iceoryx_server_abstraction) {
+          RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
+            iceoryx_server_abstraction->~IceoryxServer(), IceoryxServer)
+          rmw_free(iceoryx_server_abstraction);
+        }
+        if (rmw_service->service_name) {
+          rmw_free(const_cast<char *>(rmw_service->service_name));
+        }
+        rmw_service_free(rmw_service);
+        returnOnError = true;
+      }
+    };
 
   rmw_service = rmw_service_allocate();
   if (!rmw_service) {
     RMW_SET_ERROR_MSG("failed to allocate memory for service");
+    cleanupAfterError();
+    return nullptr;
+  }
+
+  iceoryx_server =
+    static_cast<iox::popo::UntypedServer *>(rmw_allocate(
+      sizeof(iox::popo::UntypedServer)));
+
+  if (!iceoryx_server) {
+    RMW_SET_ERROR_MSG("failed to allocate memory for iceoryx server");
+    cleanupAfterError();
+    return nullptr;
+  }
+
+  RMW_TRY_PLACEMENT_NEW(
+    iceoryx_server, iceoryx_server,
+    cleanupAfterError(), iox::popo::UntypedServer, service_description,
+    iox::popo::ServerOptions{
+      qos_policies->depth, iox::into<iox::lossy<iox::NodeName_t>>(node_full_name)});
+  if (returnOnError) {
     return nullptr;
   }
 
-  void * info = nullptr;
+  iceoryx_server->offer();
+
+  iceoryx_server_abstraction =
+    static_cast<IceoryxServer *>(rmw_allocate(sizeof(IceoryxServer)));
+  if (!iceoryx_server_abstraction) {
+    RMW_SET_ERROR_MSG("failed to allocate memory for rmw iceoryx publisher");
+    cleanupAfterError();
+    return nullptr;
+  }
+  RMW_TRY_PLACEMENT_NEW(
+    iceoryx_server_abstraction, iceoryx_server_abstraction,
+    cleanupAfterError(), IceoryxServer, type_supports, iceoryx_server);
+  if (returnOnError) {
+    return nullptr;
+  }
 
   rmw_service->implementation_identifier = rmw_get_implementation_identifier();
-  rmw_service->data = info;
+  rmw_service->data = iceoryx_server_abstraction;
 
   rmw_service->service_name =
     static_cast<const char *>(rmw_allocate(sizeof(char) * strlen(service_name) + 1));
   if (!rmw_service->service_name) {
-    RMW_SET_ERROR_MSG("failed to allocate memory for publisher topic name");
-  } else {
-    memcpy(const_cast<char *>(rmw_service->service_name), service_name, strlen(service_name) + 1);
+    RMW_SET_ERROR_MSG("failed to allocate memory for service name");
+    cleanupAfterError();
+    return nullptr;
   }
+  memcpy(const_cast<char *>(rmw_service->service_name), service_name, strlen(service_name) + 1);
```

Comment:
@elBoberido Because I'm a plumber and not a porcelain guy :clown_face: Kidding aside, this was just a copy & paste. Can be adapted with a bigger refactoring I added it to #99